_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,626 @@
+script and a version (only version 0xc0 is defined in this proposal;
+other versions are available for future soft fork upgrades).
+
+After the two scripts are hashed, their digests are put in lexicographic
+order.  This ordering will allow later verification of a merkle
+inclusion proof without knowing whether or not each leaf or branch
+appeared in the tree to the left or right of its pair sibling, thus
+reducing the amount of data that needs to be communicated as well as
+potentially improving privacy.  After ordering, the two hashes will be
+hashed together with the prefix tag "TapBranch".  As this merkle tree
+only has two leaves, the resultant hash is the merkle root.
+
+    FIXME: diagram of tree
+
+This merkle tree only covers two of the HTLC's possible end states.  For
+the third case where Alice and Bob mutually agree on a spend, they can
+use signature aggregation using something like MuSig.  As in the
+multisig section above, they work together to create a single pubkey,
+called the *Taproot internal key.*
+
+The merkle root and the internal key are then hashed together (prefix
+tag, "TapTweak") and the resultant digest is used as a private key from
+which a pubkey is derived, this value being known as the *tweak.*  The
+tweak pubkey is added to the internal key (the same operation used in
+[BIP32][] HD key derivation) in order to derive the *taproot output
+key---*the key that is used in any bech32 addresses and scriptPubKeys
+that pay these three conditions.
+
+    FIXME: diagram of tweaking, similar to one from BitMex article
+
+When it comes time to spend this money, either Alice or Bob can provide
+the script they want to use, the data needed by it (a signature and,
+sometimes, a hash preimage), the Taproot internal key, and the hash of
+the script they didn't use.  Alternatively, Alice and Bob work together
+to use signature aggregation (after accounting for the tweak) to provide
+a signature in combination with the same single-key, single-signature
+spending pattern described in the previous subsections.  As long as the
+data they provide in either case is correct, the spend will be accepted.
+
+    FIXME: table example of HTLC cost in current P2WSH and taproot for
+    each of the three possible outcomes
+
+    P2WSH
+      Output: 8+1+34
+      1.preimage, digest, A pubkey, B pubkey, OP_IF/ENDIF, Alice sig
+      2. digest, A pubkey, B pubkey, IF/ENDIF, Bob sig
+      3. N/A (use one of the above)
+
+    Taproot
+      Output: 8+1+35
+      1. preimage, digest, A pubkey, A sig; internal key; leaf hash
+      2. B pubkey, B sig; internal key; leaf hash
+      3. AB sig
+
+Because we chose to use a small tree in the above example, we will
+describe some things that would've been clearer with a larger example.
+First, if you have any unpaired nodes in your tree, your software will
+be responsible for pairing them with something, e.g a 32-byte random
+number.  Additionally, the maximum depth of a tree is 32 rows, allowing
```

Comment:
> I'd probably calculate Hash_TapLeaf("OP_RETURN OP_PUSH32 <256 bit random number>") to be sure it's unspendable.

I originally wrote that, but on reflection, it seemed to me that the probability of a digest of a random number being the preimage for a known spendable script is the same as the probability of a random number itself being the preimage for a known spendable script.  I'd love to know if I reasoned that incorrectly, though.

> Doing Hash_TapLeaf("<256bit rand>") could end up spendable

I was thinking of just using "<256bit rand>" in the branch construction so that it doesn't need to hashed (a useful optimization if you have to fill in a lot of blanks in your tree).  E.g., if 0x1234...cdef is the hash of a leaf for which I don't have a script to put in its pair sibling, I'd do `Hash_TapBranch(sort("0x1234...cdef", random_bytes(32)))`.

OTOH, having thought about it a bit more while writing this, I guess you could just fill in blank spots like that by duplicating a functional leaf you already had.  That would cut down on the storage requirements and wouldn't increase the amount of hashing necessary.  Though I guess it might be less private that way if you paired a leaf with itself.

Reply:
The advantage of creating unspendable leaves as an actual unspendable script rather than a random value is that you can prove to other participants (co-signers, ...) that it is in fact an unspendable branch.

Reply:
Ah!  Thanks!